File: sdks/sandbox/python/src/opensandbox/internal/lifecycle_metrics.py
# Copyright 2026 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Best-effort SDK metrics reporting to the lifecycle server."""

from __future__ import annotations

import asyncio
import logging
import threading
from typing import Any, Protocol

import httpx

logger = logging.getLogger(__name__)

_SDK_LANGUAGE = "python"


class _MetricsConnection(Protocol):
    headers: dict[str, str]
    user_agent: str

    def get_api_key(self) -> str: ...

    def get_base_url(self) -> str: ...

    @property
    def request_timeout(self) -> Any: ...


def _sdk_version() -> str:
    try:
        from opensandbox import __version__

        return __version__
    except Exception:
        return "0.0.0"


def _build_payload(
    *,
    sandbox_id: str | None,
    image: str | None,
    create_duration_ms: int,
    success: bool,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "eventType": "sandbox.create",
        "createDurationMs": max(0, int(create_duration_ms)),
        "sdkLanguage": _SDK_LANGUAGE,
        "sdkVersion": _sdk_version(),
        "success": bool(success),
    }
    if sandbox_id:
        payload["sandboxId"] = sandbox_id
    if image:
        payload["image"] = image
    return payload


def _headers(config: _MetricsConnection) -> dict[str, str]:
    headers = {"Content-Type": "application/json", **(config.headers or {})}
    api_key = config.get_api_key()
    if api_key and "OPEN-SANDBOX-API-KEY" not in headers:
        headers["OPEN-SANDBOX-API-KEY"] = api_key
    if config.user_agent and "User-Agent" not in headers and "user-agent" not in headers:
        headers["User-Agent"] = config.user_agent
    return headers


def _timeout_seconds(config: _MetricsConnection) -> float:
    timeout = config.request_timeout
    total = getattr(timeout, "total_seconds", None)
    if callable(total):
        return float(total())
    return float(timeout)


def _post_sync(config: _MetricsConnection, payload: dict[str, Any]) -> None:
    url = f"{config.get_base_url().rstrip('/')}/metrics/events"
    with httpx.Client(timeout=_timeout_seconds(config)) as client:
        client.post(url, json=payload, headers=_headers(config))


async def _post_async(config: _MetricsConnection, payload: dict[str, Any]) -> None:
    url = f"{config.get_base_url().rstrip('/')}/metrics/events"
    async with httpx.AsyncClient(timeout=_timeout_seconds(config)) as client:
        await client.post(url, json=payload, headers=_headers(config))


def report_sandbox_create_metric(
    config: _MetricsConnection,
    *,
    sandbox_id: str | None,
    image: str | None,
    create_duration_ms: int,
    success: bool,
) -> None:
    """Fire-and-forget create latency report. Never raises or blocks callers."""
    payload = _build_payload(
        sandbox_id=sandbox_id,
        image=image,
        create_duration_ms=create_duration_ms,
        success=success,
    )

    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        def _thread_run() -> None:
            try:
                _post_sync(config, payload)
            except Exception:
                logger.debug("Failed to report sandbox.create metrics", exc_info=True)

        threading.Thread(target=_thread_run, daemon=True).start()
        return

    async def _run() -> None:
        try:
            await _post_async(config, payload)
        except Exception:
            logger.debug("Failed to report sandbox.create metrics", exc_info=True)

    loop.create_task(_run())

File: sdks/sandbox/javascript/src/internal/lifecycleMetrics.ts
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { DEFAULT_USER_AGENT } from "../core/constants.js";
import type { ConnectionConfig } from "../config/connection.js";

const DISABLE_METRICS_ENV = "OPENSANDBOX_DISABLE_METRICS";

export interface SandboxCreateMetricsPayload {
  eventType: "sandbox.create";
  sandboxId?: string;
  image?: string;
  createDurationMs: number;
  success: boolean;
}

function readEnv(name: string): string | undefined {
  const env = (globalThis as { process?: { env?: Record<string, string | undefined> } })
    ?.process?.env;
  const v = env?.[name];
  return typeof v === "string" && v.length ? v : undefined;
}

function envMetricsDisabled(): boolean {
  return readEnv(DISABLE_METRICS_ENV)?.trim() === "1";
}

function metricsDisabled(connectionConfig: ConnectionConfig): boolean {
  return connectionConfig.disableMetrics || envMetricsDisabled();
}

/**
 * Best-effort fire-and-forget report of sandbox create latency.
 * Never throws and never awaits for callers.
 * SDK identity is conveyed via the User-Agent header.
 */
export function reportSandboxCreateMetric(
  connectionConfig: ConnectionConfig,
  opts: {
    sandboxId?: string;
    image?: string;
    createDurationMs: number;
    success: boolean;
  }
): void {
  if (metricsDisabled(connectionConfig)) {
    return;
  }

  const payload: SandboxCreateMetricsPayload = {
    eventType: "sandbox.create",
    createDurationMs: Math.max(0, Math.floor(opts.createDurationMs)),
    success: opts.success,
  };
  if (opts.sandboxId) {
    payload.sandboxId = opts.sandboxId;
  }
  if (opts.image) {
    payload.image = opts.image;
  }

  const url = `${connectionConfig.getBaseUrl().replace(/\/$/, "")}/metrics/events`;
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    ...(connectionConfig.headers ?? {}),
  };
  if (
    connectionConfig.apiKey &&
    !headers["OPEN-SANDBOX-API-KEY"]
  ) {
    headers["OPEN-SANDBOX-API-KEY"] = connectionConfig.apiKey;
  }
  if (
    !headers["User-Agent"] &&
    !headers["user-agent"]
  ) {
    headers["User-Agent"] =
      connectionConfig.userAgent || DEFAULT_USER_AGENT;
  }

  void connectionConfig
    .fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
    })
    .then(async (res) => {
      // Drain body to allow connection reuse; ignore status.
      try {
        await res.arrayBuffer();
      } catch {
        // ignore
      }
    })
    .catch(() => {
      // Metrics must never affect create UX.
    });
}
